activedata_etl/transforms/per_test_to_es.py

```diff
@@ -0,0 +1,220 @@
+#
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this file,
+# You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+# Author: Marco Castelluccio (mcastelluccio@mozilla.com)
+#
+from __future__ import division
+from __future__ import unicode_literals
+
+import json
+from zipfile import ZipFile
+
+from activedata_etl import etl2key
+from activedata_etl.imports.coverage_util import TUID_BLOCK_SIZE, download_file
+from jx_python import jx
+from mo_dots import wrap, set_default
+from mo_files import TempDirectory
+from mo_json import value2json
+from mo_logs import Log, machine_metadata
+from mo_times.dates import Date
+from mo_times.timer import Timer
+
+
+urls_w_uncoverable_lines = set()
+
+
+def process_per_test_artifact(source_key, resources, destination, task_cluster_record, artifact, artifact_etl, please_stop):
+
+    def create_record(parent_etl, count, filename, covered, uncovered):
+        file_details = resources.file_mapper.find(source_key, filename, artifact, task_cluster_record)
+
+        coverable_lines = len(covered) + len(uncovered)
+
+        if not coverable_lines and artifact.url not in urls_w_uncoverable_lines:
+            urls_w_uncoverable_lines.add(artifact.url)
+            Log.warning("per-test-coverage {{url}} has uncoverable lines", url=artifact.url)
+
+        new_record = set_default(
+            {
+                "source": {
+                    "file": set_default(
+                        file_details,
+                        {
+                            "covered": covered,
+                            "uncovered": uncovered,
+                            "total_covered": len(covered),
+                            "total_uncovered": len(uncovered),
+                            "percentage_covered": len(covered) / coverable_lines if coverable_lines else None
+                        }
+                    )
+                },
+                "etl": {
+                    "id": count(),
+                    "source": parent_etl,
+                    "type": "join",
+                    "machine": machine_metadata,
+                    "timestamp": Date.now()
+                }
+            },
+            task_cluster_record
+        )
+
+        return new_record
+
+    def process_source_file(parent_etl, count, test, sf):
+        sf = wrap(sf)
+
+        covered = []
+        uncovered = []
+        coverable = []
+        for l, c in enumerate(sf["coverage"]):
+            if c is None:
+                continue
+            coverable.append(l)
+            if c > 0:
+                covered.append(l)
+            else:
+                uncovered.append(l)
+
+        # turn covered into a set for use later
+        file_covered = set(covered)
+        coverable = set(coverable)
+
+        record = create_record(parent_etl, count, sf["name"], covered, uncovered)
+        record.test = {
+            "name": test,
```

Comment:
add a `record.test.suite` property, in case we want to aggregate coverage by suite.

Reply:
Done!